compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp

```diff
@@ -64,6 +65,179 @@ void transform_dialect::MapNestedForallToGpuThreadsOp::build(
+  if (forallOp.getNumResults() > 0)
+    return failureHelper("only bufferized scf.forall lowers to gpu.thread_id");
+  if (forallOp.getRank() > 3)
+    return failureHelper(
+        "scf.forall with rank > 3 does not lower to gpu.thread_id");
+  if (llvm::any_of(forallOp.getMixedUpperBound(), [](OpFoldResult ofr) {
+        return !getConstantIntValue(ofr).has_value();
+      })) {
+    return failureHelper("unsupported dynamic blockdim size");
+  }
+  if (!forallOp.getMapping().has_value())
+    return failureHelper("mapping must be present");
+  SmallVector<Attribute> threadMapping =
+      llvm::to_vector(forallOp.getMapping()->getValue());
+
+  // Step 1. Complete the threadMapping to a full mapping (with 1s) if
+  // necessary.
+  SmallVector<Value> numThreads = forallOp.getUpperBound(rewriter);
+  // Ensure we have 3 block sizes, one for each id.
+  Value one;
+  for (auto attr : threadMappingAttributes) {
+    if (std::find(threadMapping.begin(), threadMapping.end(), attr) ==
+        threadMapping.end()) {
+      threadMapping.push_back(attr);
+      one = one ? one : rewriter.create<arith::ConstantIndexOp>(loc, 1);
+      numThreads.push_back(one);
+    }
+  }
+
+  // Step 2. sort the values by the corresponding DeviceMappingAttrInterface.
+  auto comparator = [&](DeviceMappingAttrInterface a,
+                        DeviceMappingAttrInterface b) -> bool {
+    return a.getMappingId() < b.getMappingId();
+  };
+  SmallVector<Value> blockDimValues = scf::ForallOp::getValuesSortedByKey(
+      threadMapping, numThreads, comparator);
+  SmallVector<int64_t> blockDims =
+      llvm::to_vector(llvm::map_range(blockDimValues, [](Value v) {
+        return v.getDefiningOp<arith::ConstantIndexOp>().value();
+      }));
+
+  // Step 3. Create the gpu.thread ops and map the induction variables to the
+  // newly created ops.
+  Value flatThreadId = flatId;
+  Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+  SmallVector<Value> threadOpsUpdated;
+  int64_t flatblockDim = 1;
+  for (int64_t dim : blockDims) {
+    flatblockDim *= dim;
+  }
+  int64_t dimUsed = 1;
+  for (int64_t dim : blockDims) {
+    if (dim == 1) {
+      threadOpsUpdated.push_back(zero);
+      continue;
+    }
+    dimUsed *= dim;
+    AffineExpr d0 = rewriter.getAffineDimExpr(0);
+    Value dimId =
+        dimUsed == flatblockDim
+            ? flatThreadId
+            : makeComposedAffineApply(rewriter, loc, d0 % dim, {flatThreadId});
+    threadOpsUpdated.push_back(dimId);
+    flatThreadId = makeComposedAffineApply(rewriter, loc, d0.floorDiv(dim),
+                                           {flatThreadId});
+  }
+  IRMapping bvm;
+  for (auto [blockIdx, blockDim] :
+       llvm::zip(forallOp.getInductionVars(), threadMapping)) {
+    bvm.map(blockIdx,
+            threadOpsUpdated[blockDim.cast<DeviceMappingAttrInterface>()
+                                 .getMappingId()]);
+  }
+
+  // Step 4. Maybe create conditionals to predicate the region.
```

Comment:
I should finally be able to play with this today.

Quick offhand comment, can we refactor things upstream and avoid all the copy-pasta here and below ?

Reply:
It can go upstream if we are confident that it is what we need. I’m not fully sure this is the right logic until we plumb it through. What do you think?